crates/genie-rec/src/reader.rs

```diff
@@ -0,0 +1,314 @@
+use std::cmp::min;
+use std::io;
+use std::io::{BufRead, Read};
+
+/// A struct implementing `BufRead` that uses a small, single-use, stack-allocated buffer, intended
+/// for reading only the first few bytes from a file.
+pub(crate) struct SmallBufReader<R>
+where
+    R: Read,
+{
+    buffer: [u8; 256],
+    pointer: usize,
+    reader: R,
+}
+
+impl<R> SmallBufReader<R>
+where
+    R: Read,
+{
+    pub(crate) fn new(reader: R) -> Self {
+        Self {
+            buffer: [0; 256],
+            pointer: 0,
+            reader,
+        }
+    }
+}
+
+impl<R> Read for SmallBufReader<R>
+where
+    R: Read,
+{
+    fn read(&mut self, output: &mut [u8]) -> io::Result<usize> {
+        self.reader.read(output)
+    }
+}
+
+impl<R> BufRead for SmallBufReader<R>
+where
+    R: Read,
+{
+    fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        self.reader.read_exact(&mut self.buffer[self.pointer..])?;
+        Ok(&self.buffer[self.pointer..])
+    }
+
+    fn consume(&mut self, len: usize) {
+        self.pointer += len;
+    }
+}
+
+/// dynamically allocated buffered reader, by method of "inflation"
+/// The inflation allows us to "peek" into streams, without destroying data
+/// Reading from the buffer also only advances the pointer inside the buffer
+/// Only once another peek is done, data is rearranged (or simply overwritten if the whole buffer has been consumed)
+#[derive(Debug)]
+pub struct InflatableReader<R> {
+    /// inner reader
+    inner: R,
+    /// buffer for peek support, only inflates when needed
+    inflatable_buffer: Vec<u8>,
+    /// Position in our inflatable buffer
+    position_in_buffer: usize,
+}
+
+impl<R> InflatableReader<R> {
+    pub fn new(inner: R) -> Self {
+        Self::new_with_capacity(inner, 4096)
+    }
+
+    pub fn new_with_capacity(inner: R, capacity: usize) -> Self {
+        InflatableReader {
+            inner,
+            inflatable_buffer: Vec::with_capacity(capacity),
+            position_in_buffer: 0,
+        }
+    }
+}
+
+pub trait Peek {
+    fn peek(&mut self, amount: usize) -> io::Result<&[u8]>;
+}
+
+impl<R: Read> Peek for &mut InflatableReader<R> {
+    fn peek(&mut self, amount: usize) -> io::Result<&[u8]> {
+        (*self).peek(amount)
+    }
+}
+
+impl<R: Read> Peek for InflatableReader<R> {
+    /// Peek into inner reader, returns a slice owned by the [InflatableReader],
+    /// if data isn't available in the buffer yet, it will read it into the inner buffer
+    /// and inflate the buffer if needed.
+    fn peek(&mut self, amount: usize) -> io::Result<&[u8]> {
+        // cache this info, since we fuck around with position_in_buffer
+        let buffered_data_length = self.inflatable_buffer.len() - self.position_in_buffer;
+
+        // quick return because we have all the data to peek already
+        if buffered_data_length >= amount {
+            return Ok(
+                &self.inflatable_buffer[self.position_in_buffer..self.position_in_buffer + amount]
```

Comment:
For me to understand: Why not an inclusive range (`..=`), asking because we have `>=` in L100?

Would the following be also possible?
```suggestion
                &self.inflatable_buffer[self.position_in_buffer..=(self.position_in_buffer + amount)]
```

Reply:
talked about it on discord directly

Reply:
tl;dr;

it's about the same as `buffer[..buffer.len()]` thus using `..=` will cause an out-of-bounds